scipy/optimize/_lsq/least_squares.py

```diff
@@ -0,0 +1,900 @@
+        return
+    rho[1, mask] = 1
+    rho[1, ~mask] = z[~mask]**-0.5
+    rho[2, mask] = 0
+    rho[2, ~mask] = -0.5 * z[~mask]**-1.5
+
+
+def soft_l1(z, rho, cost_only):
+    t = 1 + z
+    rho[0] = 2 * (t**0.5 - 1)
+    if cost_only:
+        return
+    rho[1] = t**-0.5
+    rho[2] = -0.5 * t**-1.5
+
+
+def cauchy(z, rho, cost_only):
+    rho[0] = np.log1p(z)
+    if cost_only:
+        return
+    t = 1 + z
+    rho[1] = 1 / t
+    rho[2] = -1 / t**2
+
+
+def arctan(z, rho, cost_only):
+    rho[0] = np.arctan(z)
+    if cost_only:
+        return
+    t = 1 + z**2
+    rho[1] = 1 / t
+    rho[2] = -2 * z / t**2
+
+
+IMPLEMENTED_LOSSES = dict(linear=None, huber=huber, soft_l1=soft_l1,
+                          cauchy=cauchy, arctan=arctan)
+
+
+def construct_loss_function(m, loss, loss_scale):
+    if loss == 'linear':
+        return None
+
+    if not callable(loss):
+        loss = IMPLEMENTED_LOSSES[loss]
+        rho = np.empty((3, m))
+
+        def loss_function(f, cost_only=False):
+            z = (f / loss_scale) ** 2
+            loss(z, rho, cost_only=cost_only)
+            if cost_only:
+                return 0.5 * loss_scale**2 * np.sum(rho[0])
+            rho[0] *= loss_scale**2
+            rho[2] /= loss_scale**2
+            return rho
+    else:
+        def loss_function(f, cost_only=False):
+            z = (f / loss_scale) ** 2
+            rho = loss(z)
+            if cost_only:
+                return 0.5 * loss_scale**2 * np.sum(rho[0])
+            rho[0] *= loss_scale**2
+            rho[2] /= loss_scale**2
+            return rho
+
+    return loss_function
+
+
+def least_squares(
+        fun, x0, jac='2-point', bounds=(-np.inf, np.inf), method='trf',
+        ftol=EPS**0.5, xtol=EPS**0.5, gtol=EPS**0.5, loss='linear',
+        loss_scale=1.0, scaling=1.0, diff_step=None, tr_solver=None,
+        tr_options={}, jac_sparsity=None, max_nfev=None, verbose=0, args=(),
+        kwargs={}):
+    """Minimize the robustified sum of squares of nonlinear functions, subject
+    to bound constraints on independent variables.
+
+    Let f(x) be an m-dimensional function of n variables and rho(s) be a
+    scalar function, `least_squares` finds a local minimum of::
+
+        F(x) = 0.5 * sum(rho(f_i(x)**2), i = 1, ..., m), lb <= x <= ub
+
+    We call f(x) as a vector of residuals or simply residuals, and F(x) as a
+    cost function or simply cost.
+
+    We call rho(s) as a loss function, its purpose to reduce the influence
+    of outliers on the solution.
+
+    Partial derivatives of f with respect to x form m-by-n matrix called
+    Jacobian, where an element (i, j) equals the partial derivative of f[i]
+    with respect to x[j].
+
+    Parameters
+    ----------
+    fun : callable
+        Function which computes the vector of residuals with the signature
+        ``fun(x, *args, **kwargs)``, i.e., the minimization proceeds with
+        respect to it's first argument. The argument ``x`` passed to this
```

Comment:
Any corrections to the text would be helpful. Likely there are more mistakes and generally bad usage of English.